src/shared-validator.cc

```diff
@@ -1192,6 +1210,16 @@ Result SharedValidator::OnReturnCallIndirect(const Location& loc,
   return result;
 }
 
+Result SharedValidator::OnReturnCallRef(const Location& loc,
+                                        Var function_type_var) {
+  Result result = CheckInstr(Opcode::ReturnCallRef, loc);
+  FuncType func_type;
+  result |= typechecker_.OnReturnCallRef(function_type_var.to_type());
+  result |= CheckFuncTypeIndex(function_type_var, &func_type);
+  result |= typechecker_.OnReturnCall(func_type.params, func_type.results);
```

Comment:
This code is based on the original implementation in 9aac0a28896a714b698ab489a473b9bc05deb7c3, although OnReturnCallIndirect uses its own function which report errors as "call_indirect" rather than "call". Maybe adding a string "descriptor" argument would be better than duplicating the whole function.

Reply:
Seems like a good idea, could certainly be in a followup too